### PR TITLE
Update Storage Tests

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Updates the versions of `spock-core` being used in Storage testing to the latest release as it introduces better compatibilities with Java 17. Began leveraging newer features from `azure-core-test` to enable the usage annotation such as `DoNotRecord` to allow for the creation of unrecorded tests within test classes that record network calls (for scenarios that test non-network calls such as invalid parameters being passed).

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareAsyncClient.java [hunk 1]
```diff
@@ -411,12 +411,12 @@ public Mono<ShareInfo> createIfNotExists() {
...
-     *     &#125; else &#123;
-     *         System.out.println&#40;&quot;successfully created.&quot;&#41;;
-     *     &#125;
-     * &#125;&#41;;
+     *         if &#40;response.getStatusCode&#40;&#41; == 409&#41; &#123;
+     *             System.out.println&#40;&quot;Already exists.&quot;&#41;;
+     *         &#125; else &#123;
+     *             System.out.println&#40;&quot;successfully created.&quot;&#41;;
+     *         &#125;
+     *     &#125;&#41;;
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareAsyncClient.java:419]: nit: we should avoid unrelated formatting changes.

File: sdk/storage/azure-storage-common/src/test-shared/java/com/azure/storage/common/test/shared/StorageSpec.groovy [hunk 2]
```diff
@@ -62,7 +67,7 @@ class StorageSpec extends Specification {
         return TestDataFactory.getInstance();
     }
 
-    protected <T> T instrument(T builder) {
+    protected <T extends HttpTrait<T>> T instrument(T builder) {
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-common/src/test-shared/java/com/azure/storage/common/test/shared/StorageSpec.groovy:70]: nice, can we go ahead and refactor the "groovyReflectiveCallBelow"()?

**Review by kasobol-msft — COMMENTED** [review]
LGTM.

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareAsyncClient.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareAsyncClient.java:419]: This came from the `maven-codesnippet-plugin` updating outdated code samples. I'll need to work on an update to exclude it conditionally in local builds.

Comment (kasobol-msft) [reply sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareAsyncClient.java:419]: best if we can harden analyzers to enforce one formatting style here and in the code.
it seems that people commit code formatted by whatever is in their local settings and then we have PRs that contain a lot of noise.

**Review by kasobol-msft — APPROVED** [review]

**Comment by alzimmermsft** [comment]
Equivalent, or better, change was already done in another PR
